File: core/src/pl/shockah/godwit/geom/IVec2.java
package pl.shockah.godwit.geom;

import javax.annotation.Nonnull;

import pl.shockah.unicorn.ease.Easable;
import pl.shockah.unicorn.ease.Easing;

public abstract class IVec2 implements Easable<IVec2> {
	IVec2() {
	}

	@Override
	public boolean equals(Object o) {
		if (!(o instanceof IVec2))
			return false;
		IVec2 v = (IVec2)o;
		return x() == v.x() && y() == v.y();
	}

	@Override
	public int hashCode() {
		return Float.floatToIntBits(x()) * 31 + Float.floatToIntBits(y());
	}

	public abstract float x();
	public abstract float y();

	@Nonnull
	public abstract IVec2 getCopy();

	@Nonnull
	public Vec2 add(float x, float y) {
		return x == 0 && y == 0 ? asImmutable() : new Vec2(x() + x, y() + y);
	}

	@Nonnull
	public final Vec2 add(@Nonnull IVec2 v) {
		return add(v.x(), v.y());
	}

	@Nonnull
	public Vec2 subtract(float x, float y) {
		return x == 0 && y == 0 ? asImmutable() : new Vec2(x() - x, y() - y);
	}

	@Nonnull
	public final Vec2 subtract(@Nonnull IVec2 v) {
		return subtract(v.x(), v.y());
	}

	@Nonnull
	public Vec2 multiply(float x, float y) {
		return x == 1 && y == 1 ? asImmutable() : new Vec2(x() * x, y() * y);
	}

	@Nonnull
	public final Vec2 multiply(@Nonnull IVec2 v) {
		return multiply(v.x(), v.y());
	}

	@Nonnull
	public final Vec2 multiply(float f) {
		return multiply(f, f);
	}

	@Nonnull
	public Vec2 divide(float x, float y) {
		return x == 1 && y == 1 ? asImmutable() : new Vec2(x() / x, y() / y);
	}

	@Nonnull
	public final Vec2 divide(@Nonnull IVec2 v) {
		return divide(v.x(), v.y());
	}

	@Nonnull
	public final Vec2 divide(float f) {
		return divide(f, f);
	}

	@Nonnull
	public Vec2 negate() {
		return x() == 0 && y() == 0 ? asImmutable() : new Vec2(-x(), -y());
	}

	@Nonnull
	public Vec2 withX(float x) {
		return x == x() ? asImmutable() : new Vec2(x, y());
	}

	@Nonnull
	public Vec2 withY(float y) {
		return y == y() ? asImmutable() : new Vec2(x(), y);
	}

	@Nonnull
	public final Vec2 getAbs() {
		return new Vec2(Math.abs(x()), Math.abs(y()));
	}

	public final float getLength() {
		return (float)Math.sqrt(x() * x() + y() * y());
	}

	public final float getAngle() {
		return Vec2.zero.getAngle(this);
	}

	public final float getAngle(@Nonnull IVec2 v) {
		return (float)Math.toDegrees(Math.atan2(y() - v.y(), v.x() - x()));
	}

	@Nonnull
	public final Vec2 getNormalized() {
		float length = getLength();
		if (length == 0f || length == 1f)
			return asImmutable();
		else
			return this.multiply(1f / length);
	}

	@Nonnull
	public final MutableVec2 mutableCopy() {
		return new MutableVec2(x(), y());
	}

	@Nonnull
	public Vec2 asImmutable() {
		return new Vec2(x(), y());
	}

	@Override
	@Nonnull
	public final Vec2 ease(@Nonnull IVec2 other, float f) {
		return new Vec2(Easing.linear.ease(x(), other.x(), f), Easing.linear.ease(y(), other.y(), f));
	}
}

File: core/src/pl/shockah/godwit/geom/IVec2.kt
package pl.shockah.godwit.geom

import com.badlogic.gdx.math.Vector2
import pl.shockah.godwit.ease.Easable
import kotlin.math.atan2
import kotlin.math.sqrt

typealias Vec2 = IVec2<*>

abstract class IVec2<T : IVec2<T>> : Comparable<IVec2<*>>, Easable<IVec2<*>> {
	abstract val x: Float
	abstract val y: Float

	open val lengthSquared: Float
		get() = x * x + y * y

	open val length: Float
		get() = sqrt(lengthSquared)

	open val degrees: Degrees
		get() = ImmutableVec2.ZERO degrees this

	open val radians: Radians
		get() = ImmutableVec2.ZERO radians this

	open val gdx: Vector2
		get() = Vector2(x, y)

	open val normalized: T
		get() {
			val length = length
			@Suppress("UNCHECKED_CAST")
			return when (length) {
				0f -> throw IllegalStateException()
				1f -> this as T
				else -> this / length
			}
		}

	operator fun get(index: Int): Float {
		return when (index) {
			0 -> x
			1 -> y
			else -> throw IllegalArgumentException()
		}
	}

	operator fun component1(): Float {
		return x
	}

	operator fun component2(): Float {
		return y
	}

	open fun immutable(): ImmutableVec2 {
		return ImmutableVec2(x, y)
	}

	fun mutableCopy(): MutableVec2 {
		return MutableVec2(x, y)
	}

	abstract operator fun unaryMinus(): T

	abstract operator fun plus(vector: Vec2): T

	abstract operator fun minus(vector: Vec2): T

	abstract operator fun times(vector: Vec2): T

	abstract operator fun div(vector: Vec2): T

	abstract operator fun plus(scalar: Float): T

	abstract operator fun minus(scalar: Float): T

	abstract operator fun times(scalar: Float): T

	abstract operator fun div(scalar: Float): T

	infix fun degrees(vector: Vec2): Degrees {
		return (this radians vector).degrees
	}

	infix fun radians(vector: Vec2): Radians {
		return atan2((y - vector.y).toDouble(), (vector.x - x).toDouble()).toFloat().radians
	}

	infix fun dot(vector: Vec2): Float {
		return x * vector.x + y * vector.y
	}

	infix fun cross(vector: Vec2): Float {
		return x * vector.y - y * vector.x
	}

	abstract infix fun rotated(angle: Angle): T

	infix fun equals(other: Vec2): Boolean {
		return other.x == x && other.y == y
	}

	infix fun notEquals(other: Vec2): Boolean {
		return other.x != x || other.y != y
	}

	override fun equals(other: Any?): Boolean {
		return other is Vec2 && other.x == x && other.y == y
	}

	override fun hashCode(): Int {
		return x.hashCode() * 31 + y.hashCode()
	}

	override fun toString(): String = "[$x, $y]"

	override fun compareTo(other: IVec2<*>): Int {
		return length.compareTo(other.length)
	}

	override fun ease(other: IVec2<*>, f: Float): IVec2<*> {
		return ImmutableVec2(
				x.ease(other.x, f),
				y.ease(other.y, f)
		)
	}
}